### PR TITLE
[PyTorch] Support L2Norm basic op and module -> use for qk_norm

### PR DESCRIPTION
**Pull request by negvet**
# Description

Support `L2Norm` basic op and module.
Enable it for qk normalization via `use_qk_norm`.
Exposure to MHA and transformer block modules.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Support `L2Norm` basic op and module
- Support qk_norm

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/module/l2norm.py
Comment (timmoon10) [inline transformer_engine/pytorch/module/l2norm.py:1]: Do we need to have a separate implementation here? We have separate LayerNorm and RMSNorm modules for backward compatibility, so no design reason to copy that here.

File: transformer_engine/pytorch/module/l2norm.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        # JIT warmup for L2Norm fused operations
+        if seq_length and micro_batch_size:
+            device = getattr(self, "device", torch.device("cuda"))
+            if hasattr(device, "type") and device.type == "cuda":
+                set_jit_fusion_options()
+                # For L2Norm, we don't know the hidden size until forward pass,
+                # but we can warm up with common sizes used in attention mechanisms
+                common_hidden_sizes = [768, 1024, 2048, 4096, 8192]
+                for hidden_size in common_hidden_sizes:
+                    warmup_jit_l2norm_all_dtypes(hidden_size, seq_length, micro_batch_size)
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/l2norm.py]: Why not JIT lazily during the first forward?

File: transformer_engine/pytorch/ops/basic/l2norm.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    canonicalize_device,
+    canonicalize_dtype,
+    clear_tensor_data,
+)
+from ..op import BasicOperation, OperationContext
+from .._common import maybe_autocast_dtype, reshape
+from ...jit import l2norm_fused, l2norm_fwd_fused, l2norm_backward_fused
+
+
+class L2Norm(BasicOperation):
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/l2norm.py]: This name makes me think you are actually computing the L2 norm, like [`torch.linalg.norm`](https://docs.pytorch.org/docs/stable/generated/torch.linalg.norm.html#torch.linalg.norm). What do other implementations call it? If there are no prominent implementations, we could consider calling it `L2Normalization` to avoid being ambiguous.

File: transformer_engine/pytorch/ops/basic/l2norm.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        dtype: Optional[torch.dtype] = None,
+    ) -> None:
+        super().__init__()
+        self.eps: float = eps
+        self.device = canonicalize_device(device)
+        self.dtype = canonicalize_dtype(dtype)
+
+    def reset_parameters(self) -> None:
+        """L2Norm has no parameters to reset"""
+
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/l2norm.py]: Not needed:
```suggestion
```

File: transformer_engine/pytorch/ops/basic/l2norm.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        self,
+        *,
+        eps: float = 1e-6,
+        device: Optional[torch.device | str] = None,
+        dtype: Optional[torch.dtype] = None,
+    ) -> None:
+        super().__init__()
+        self.eps: float = eps
+        self.device = canonicalize_device(device)
+        self.dtype = canonicalize_dtype(dtype)
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/l2norm.py]: Do we need this? For LayerNorm/RMSNorm, we need to coerce the data so they are compatible with cuDNN or our custom kernels. However, `torch.compile` should be more general and should accommodate CPU data or mixed dtypes.

File: transformer_engine/pytorch/ops/basic/l2norm.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        device = self.device
+        if device.type != "cuda":
+            device = canonicalize_device(None)
+        dtype = maybe_autocast_dtype(default_dtype=self.dtype or input_.dtype)
+
+        # Reshape input for computation
+        input_dims = tuple(input_.size())
+        inner_dim = input_dims[-1]
+        x = reshape(input_, (-1, inner_dim), device=device, dtype=dtype)
+
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/l2norm.py]: For LayerNorm/RMSNorm, we reshape to 2D due to limitations in our custom kernels. However, the `torch.compile` implementation can gracefully handle multi-dimensional tensors.
```suggestion
        x = input_
```

File: transformer_engine/pytorch/ops/basic/l2norm.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        # Check if output is quantized
+        output_quantizer = None
+        if (
+            FP8GlobalStateManager.is_fp8_enabled()
+            and next_op is not None
+            and next_op.num_quantizers("forward") > 0
+        ):
+            output_quantizer = next_op.get_quantizer("forward", 0)
+
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/l2norm.py]: There's no benefit from quantizing early since we don't have a fused kernel.
```suggestion
```

File: tests/pytorch/test_fusible_ops.py [hunk 7]
```diff
@@ -1273,6 +1273,76 @@ def test_rmsnorm(
         torch.testing.assert_close(dx_test, x_ref.grad, **tols)
         torch.testing.assert_close(dw_test, w_ref.grad, **tols)
 
+    @pytest.mark.parametrize("in_shape", ((32,), (6, 16, 64), (32, 64)))
+    @pytest.mark.parametrize("dtype", _dtypes)
+    @pytest.mark.parametrize("quantization", (None, "fp8", "mxfp8"))
```
Comment (timmoon10) [inline tests/pytorch/test_fusible_ops.py]: We don't have a fused kernel, so there's not much value in testing quantization.
```suggestion
```

File: tests/pytorch/test_fusible_ops.py [hunk 8]
```diff
@@ -1273,6 +1273,76 @@ def test_rmsnorm(
...
+        # Expected numerical error
+        tols = dtype_tols(dtype)
+        if quantized_compute:
+            tols = dtype_tols(tex.DType.kFloat8E4M3)
+        else:
+            # L2Norm requires slightly looser tolerances
+            if dtype == torch.float16:
+                tols = dict(rtol=5e-3, atol=5e-4)
+            elif dtype == torch.bfloat16:
+                tols = dict(rtol=2e-2, atol=5e-3)
```
Comment (timmoon10) [inline tests/pytorch/test_fusible_ops.py]: Why? Perhaps sqrt would hurt accuracy, but why don't we see problems with LayerNorm or RMSNorm?

File: transformer_engine/pytorch/module/l2norm.py (hunk 1 above)
Comment (negvet) [reply transformer_engine/pytorch/module/l2norm.py]: This is to achieve consistency of activation recomputation. If only compile during the first forward pass, the JIT compiler might generate different kernels when the same operation is called later during activation recomputation. 

@ksivaman could you please take a look and confirm that this issue will be solved with the current approach?

File: transformer_engine/pytorch/module/l2norm.py
Comment (negvet) [reply transformer_engine/pytorch/module/l2norm.py:1]: Got it, removed this module implementation in https://github.com/NVIDIA/TransformerEngine/pull/1864/commits/aabeea28dd3338aff22b329c1d20fea85db6d90d

File: transformer_engine/pytorch/ops/basic/l2norm.py (hunk 2 above)
Comment (negvet) [reply transformer_engine/pytorch/ops/basic/l2norm.py]: Good point. My naming followed the one from llama4 in transformers - `Llama4TextL2Norm`.
But I agree that it sounds like a norm computation. In addition, we have `transformer_engine/common/multi_tensor/l2norm.cu`, which is a L2 norm computation. 
Renamed to `L2Normalization`.

File: transformer_engine/pytorch/ops/basic/l2norm.py (hunk 4 above)
Comment (negvet) [reply transformer_engine/pytorch/ops/basic/l2norm.py]: I see, removed it

File: transformer_engine/pytorch/ops/basic/l2norm.py (hunk 5 above)
Comment (negvet) [reply transformer_engine/pytorch/ops/basic/l2norm.py]: Fixed

File: transformer_engine/pytorch/ops/basic/l2norm.py (hunk 6 above)
Comment (negvet) [reply transformer_engine/pytorch/ops/basic/l2norm.py]: Fixed

File: tests/pytorch/test_fusible_ops.py (hunk 7 above)
Comment (negvet) [reply tests/pytorch/test_fusible_ops.py]: Fixed!

File: tests/pytorch/test_fusible_ops.py (hunk 8 above)
Comment (negvet) [reply tests/pytorch/test_fusible_ops.py]: Thanks for noticing.
I have narrowed it down, and it turns out that only for bf16 backprop I need to increase atol from 1.5e-3 to 2e-3.
It is hard to say for sure. sqrt should probably not be the reason for that. I would say that implementation quality might be the issue for jit-compiled generic operations. Probably intermediate bf16 precision can lead to that as well.
